### ISSUE TITLE
twisted.web.util.formatFailure vulnerable to XSS in rare cases

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4896]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#4896|
|Type|defect|
|Created|2011-02-18 01:44:03Z|
|Branch|https://github.com/twisted/twisted/tree/template-failure-formatting-4896-2|

`twisted.web.util.formatFailure` assumes that locals and instance variables do not contain HTML, but that isn't necessarily the case; for example, put this in a `render_GET`:

```
locals()['<script>alert("hi")</script>'] = 1
1/0
```

(if you have `site.displayTracebacks` enabled, you'll see an alert pop up on the error page.)

Attachments:

* [4896-001-escape-everything.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e80/e80203b40a74923603cda969ff399f7f01637723/c1f23771a008aef79bcaa2f178f8460d09087dba.patch) (1304 bytes) - added by ivank on 2011-02-18 01:44:56Z - no unit tests here

<details><summary>Searchable metadata</summary>

```
trac-id__4896 4896
type__defect defect
reporter__ivank ivank
priority__normal normal
milestone__ 
branch__branches_template_failure_formatting_4896_2 branches/template-failure-formatting-4896-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__security security
time__1297993443000000 1297993443000000
changetime__1328215600000000 1328215600000000
version__None None
owner__exarkun exarkun
cc__jknight cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|ivank commented|
|-|-|

While the `locals()` example is very contrived, it's possible that someone is putting arbitrary user input into object attributes via `setattr`.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_4"></a>|@washort commented|
|-|-|

Now that we have `twisted.web.template` we should use it in `formatFailure` and `Request.processingFailed`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [33212]) Branching to 'template-failure-formatting-4896'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I re-implemented `formatFailure` in terms of `twisted.web.template`.  I included a template that roughly matches the old markup.

I dropped the rendering of locals and globals, because as of [#5011](https://github.com/twisted/twisted/issues/5011), `twisted.web` no longer passes any failures to `formatFailure` that have any locals or globals in them.  Perhaps we will want to re-introduce this functionality somehow.

I deprecated all of the support functions in `twisted.web.util` that are no longer used.  I also added a bunch of new support classes; possibly these would be better off private, or in another module.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/template-failure-formatting-4896)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Oops.  There appears to be some overlap between this ticket and [#4948](https://github.com/twisted/twisted/issues/4948).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_8"></a>|@jonathanj set owner to @exarkun|
|-|-|

Firstly, great job on this! Secondly, interpolating strings into HTML is nasty, so hooray for this change!

Some things:

 1. I notice you're using the `@` decorator syntax, at one point this was forbidden (although the coding standard doesn't currently say anything about decorators.) Was this because they weren't supported in Python 2.4 and support for that was recently dropped?

 2. Why not `yield SourceLineElement(...)` in `SourceFragmentElement.sourceLines` instead of building the list and returning it?

 3. `formatFailure` calls `flattenString` which is apparently an asynchronous API, it returns a `Deferred` anyway, but `formatFailure` is written in a synchronous fashion and makes some assumptions about the implementation of `flattenString`. What's going on here? I realise that the API for `formatFailure` was previously a synchronous one but the current code strikes me as pretty weird.

 4. I think that the new support classes should exist in a new private module.

 5. The CSS in `failure.xhtml` uses `monotype`, which is not a thing, when I think it should actually use `monospace`.

It's unfortunate that without patterns 20 lines of markup explodes into around 200 lines of supporting code and potentially diminished readability, although I think the tests do benefit.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

Thanks for the review!

```
>  1. I notice you're using the @ decorator syntax, at one point this was forbidden (although the coding standard doesn't currently say anything about decorators.) Was this because they weren't supported in Python 2.4 and support for that was recently dropped? 
```

Yea.  With Python 2.4 support out the window, there is no longer any compatibility reason to avoid using the syntax.  I don't think we need to pour over the codebase adding uses of it to every possible site, but using it in new code like this seems like the right idea.

```
>  2. Why not yield SourceLineElement(...) in SourceFragmentElement.sourceLines instead of building the list and returning it? 
```

Hmm.  I don't remember.  I changed it to a generator.

```
>  3. formatFailure calls flattenString which is apparently an asynchronous API, it returns a Deferred anyway, but formatFailure is written in a synchronous fashion and makes some assumptions about the implementation of flattenString. What's going on here? I realise that the API for formatFailure was previously a synchronous one but the current code strikes me as pretty weird. 
```

It's true, this is a bit tricky.  Since I know there are no Deferreds in the document being flattened, I know that `flattenString` is going to have finished by the time it returns.  I wouldn't want to write a lot of code that makes this assumption, but it seems to be more or less the only way to preserve the existing synchronous interface.

```
>  4. I think that the new support classes should exist in a new private module. 
```

I dunno if they need a whole module.  I can make them private in `twisted.python.util` though.  I'd like to leave `FailureElement` public, though, and encourage people to use it instead of `formatFailure`.

```
>  5. The CSS in failure.xhtml uses monotype, which is not a thing, when I think it should actually use monospace. 
```

Okay, changed.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_10"></a>|@jonathanj set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_9):

> Hmm.  I don't remember.  I changed it to a generator.

I thought it might have something to do with the tests, but those all still pass.

> It's true, this is a bit tricky.  Since I know there are no Deferreds in the document being flattened, I know that `flattenString` is going to have finished by the time it returns.  I wouldn't want to write a lot of code that makes this assumption, but it seems to be more or less the only way to preserve the existing synchronous interface.

Perhaps then there should be a test that fails when `flattenString` doesn't operate synchronously to ensure that `formatFailure` does not accidentally get broken.

> I dunno if they need a whole module.  I can make them private in `twisted.python.util` though.  I'd like to leave `FailureElement` public, though, and encourage people to use it instead of `formatFailure`.

I just thought it would be easier than adding `_` to everything everywhere, but it looks like you already did that. :)

I think this looks good to merge, please go ahead.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [33498]) Branching to 'template-failure-formatting-4896-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [33500]) Bump the version number on the new deprecation, since we missed 12.0.  Also jam a @since onto the new public element.

refs [#4896](https://github.com/twisted/twisted/issues/4896)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [33502]) Return the test Deferreds, or else.

refs [#4896](https://github.com/twisted/twisted/issues/4896)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [33503]) Merge template-failure-formatting-4896-2

Author: exarkun
Reviewer: jonathanj
Fixes: [#4896](https://github.com/twisted/twisted/issues/4896)

Change `twisted.web.util.formatFailure`, an API for rendering a `Failure` to an HTML
string, to use `twisted.web.template`.  This simplifies the code and fixes a number of
potential security issues caused by insufficient quoting in the previous implementation.
